sdk/ai/Azure.AI.Projects/src/Custom/Datasets/Datasets.cs

```diff
@@ -97,35 +114,119 @@ public Response UploadFolder(string name, string version, string folderPath)
                 throw new ArgumentException("The provided folder is empty.");
             }
 
-            RequestContent content = RequestContent.Create(new FolderDatasetVersion(dataUri: containerClient.Uri.AbsoluteUri));
-            return CreateOrUpdate(
-                name,
-                outputVersion,
-                content
-            );
+            var uriBuilder = new UriBuilder(blobClient!.Uri) { Query = "" };
+            var dataUri = uriBuilder.Uri.AbsoluteUri;
+
+            RequestContent content = new FolderDatasetVersion(dataUri: dataUri).ToRequestContent();
+            CreateOrUpdate(name, outputVersion, content);
+
+            return GetDataset(name, outputVersion);
+        }
+
+        /// <summary>
+        /// Internal helper method to create a new dataset and return a BlobContainerClient to the dataset's blob storage.
+        /// </summary>
+        private async Task<(BlobContainerClient ContainerClient, string OutputVersion)> CreateDatasetAndGetContainerClientAsync(string name, string inputVersion, string? connectionName = null)
+        {
+            var pendingUploadRequest = new PendingUploadRequest(
+                pendingUploadId: null,
+                connectionName: connectionName,
+                pendingUploadType: PendingUploadType.BlobReference,
+                serializedAdditionalRawData: null);
+
+            var pendingUploadResponse = await PendingUploadAsync(
```

Comment:
nit: We can avoid unwrapping: `pendingUploadResponse.Value` if we will explicitly mark that we need to get `PendingUploadResponse`
```C#
PendingUploadResponse pendingUploadResponse = await PendingUploadAsync(
```